txtChange Email Verification

jesma14 opened the pull request:
This pull request configures an email validation flow for new registering users.

**Core Features:**
- Email verification is sent to user email when they first register
- Verification link is set to expire in about 5 minutes after sending. (In my demo included, I've set it to a few seconds just for that purpose)
- Once user clicks the link, they are redirected to a page that tells them to continue to sign in
- Users are automatically redirected to home screen after verification is complete
- If verification link expires or it fails in some way, the user information is removed from the database
- Password validation now requires the following:
   - At least one uppercase character
   - At least one lowercase character
   - At least one digit
   - At least one special symbol
   - Must have a length between 6 and 64 characters inclusive
 - These requirements are only required for registering users. The password validation is not required for returning users, as this requirement is already configured for new users and would therefore be redundant

**BUGS**
- As part of the verification process, the user is removed from the database when the verification fails as the verification email must be sent once the user completes the registration form. In the off chance that the user completely closes the app (and the instance is destroyed) before the verification process is completed, they technically still exist in Firebase Authentication. They can't log in, and they can't access the app, but they are technically a user.

Please see this video to view the changes.

Jrod7938 commented:
Please update the email timer to be 15 minutes

Jrod7938 commented:
I have confirmed email creation works, but signing in does not.

Jrod7938 commented:
I confirmed that account creation works but logging in does not work

jesma14 commented:
@Jrod7938 can you tell me what error you get when you log in? Does it throw an error?

Jrod7938 commented:
> @Jrod7938 can you tell me what error you get when you log in? Does it throw an error?

It says invalid login details are incorrect even though I double checked that they are. Here is the error logged into LOGCAT:

Initial task failed for action RecaptchaAction(action=signInWithPassword)with exception - An internal error has occurred. [ INVALID_LOGIN_CREDENTIALS ]

jesma14 commented:
> > @Jrod7938 can you tell me what error you get when you log in? Does it throw an error?
> 
> It says invalid login details are incorrect even though I double checked that they are. Here is the error logged into LOGCAT:
> 
> Initial task failed for action RecaptchaAction(action=signInWithPassword)with exception - An internal error has occurred. [ INVALID_LOGIN_CREDENTIALS ]

Yeah, ik what the problem is. I added a check to see if the user trying to log in had a display name, due to the bug I mentioned in the description of my PR but it was a poor implementation. I just fixed it. It should work now please try again.

Jrod7938 commented:
> > > @Jrod7938 can you tell me what error you get when you log in? Does it throw an error?
> > 
> > 
> > It says invalid login details are incorrect even though I double checked that they are. Here is the error logged into LOGCAT:
> > Initial task failed for action RecaptchaAction(action=signInWithPassword)with exception - An internal error has occurred. [ INVALID_LOGIN_CREDENTIALS ]
> 
> Yeah, ik what the problem is. I added a check to see if the user trying to log in had a display name, due to the bug I mentioned in the description of my PR but it was a poor implementation. I just fixed it. It should work now please try again.

Please convert this pull request from "Draft: to "Ready for Review" so I can merge

jesma14 commented on app/src/main/java/com/jrod7938/textchangeapp/components/Components.kt:
What is this regex? Why is there an X?

Jrod7938 replied:
isbns that are 10 characters long also include letters, so users were unable to create book listings if they had an isbn 10

jesma14 replied:
Have you tested this? I haven't been able to find an ISBN with an X in it. Also, this regex implies that the ISBN can have an X anywhere in the string. Is this what you intended?

Jrod7938 replied:
Look at book: 110715104X

Jrod7938 commented on app/src/main/java/com/jrod7938/textchangeapp/screens/login/LoginScreenViewModel.kt:
```diff
@@ -89,13 +89,31 @@ class LoginScreenViewModel: ViewModel() {
                     if (task.isSuccessful){
                         Log.d("Firebase", "signInWithEmailAndPassword: ${task.result}")
-                        home()
+                        FirebaseFirestore.getInstance().collection("users")
```
this is unnecessary, the task of logging the user in has already been complete therefore, all you have to do is navigate to the Home Screen. No need to make extra firebase calls

Jrod7938 replied:
I fixed this in #948b1f4

jesma14 replied:
Please read my PR. There is a bug. A user can bypass verification by closing the app. When they reopen it, they log in with their credentials and have access. The app however, loads incorrectly because the user has no information, just an email and a password. This is bc the rest of the user's full details is written into the database after successful verification, but in this case verification hasn't been completed. To avoid this, I check to see if the user attempting to log in has a full user profile in the database. In this case, I check if they have a display name. If they do, log in.

jesma14 replied:
Here is a video of what happens in this case: https://drive.google.com/file/d/1GLyQPUSfpMKnI9GvbGUnRMx1MmWL8PEE/view?usp=sharing

jesma14 replied:
This is what happens when the unauthorized user tries to actually use the app: https://drive.google.com/file/d/1GUJICDJ_l8ESsmWCy-mmwmY8nOZIhz06/view?usp=sharing

Jrod7938 replied:
issue is fixed in 25d17aa

jesma14 replied:
No this doesn't work. In this case, it's not that ther verification fails. It's bypassed altogether. So it never reaches this step and the user is not deleted.

Jrod7938 replied:
I don't think this issue can be fixed then. The delete user function runs on the users device so if they cancel the app it never runs.

jesma14 replied:
Yes. I did fix it. It probably isn't the best implementation, but it did work. Thats why I did that second data base check.

Jrod7938 replied:
I see, I pushed your update again

Jrod7938 replied:
splash screen now checks if the Email is verified, if it is go home else go to login screen
